language-plutus-core/src/Language/PlutusCore/PrettyCfg.hs

```diff
@@ -0,0 +1,45 @@
+-- | This module contains the 'PrettyCfg' typeclass, a more sophisticated
+-- typeclass for pretty-printing that allows us to dump debug information only
+-- when wanted.
```

Comment:
`prettyprinter` allows adding annotations (adding: http://hackage.haskell.org/package/prettyprinter-1.2.1/docs/Data-Text-Prettyprint-Doc.html#g:14, rendering: http://hackage.haskell.org/package/prettyprinter-1.2.1/docs/Data-Text-Prettyprint-Doc-Render-Util-SimpleDocTree.html#v:renderSimplyDecorated), why not use that? Then we can benefit from its more sophisticated machinery too.

Reply:
I'm sure how exactly one would do that to keep type annotations as well as `Unique` names.

Reply:
Not sure I got the idea across: at the moment afaict all the debug printing does is add the unique numbers as suffixes. So what we could do instead is always print them, but annotate them with a `Debug` flag. Then you can tell the printing function to either include them or not include them.

Reply:
I think you would need another typeclass, though, since you need to specify the type of the annotation. So something like
```
data DocAnn = Debug
class PrettyA a where
    prettyA :: a -> Doc DocAnn
instance PrettyA (Name a) where
    prettyA (Name _ s (Unique u)) = (pretty (decodeUtf8 (BSL.toStrict s))) <> (PP.annotate Debug $ "_" <> pretty u)
```
then you can use `renderSimplyDecorated` to ignore `Debug` annotated sections or not. I think this is somewhat nicer since the configuration will only appear at the printing stage, and the pretty-printers can just unconditionally stick stuff in with appropriate flags.

Anyway, not sure I feel very strongly about this.

Reply:
Right, that was where I was confused. `pretty` has to be of type `a -> Doc b`. So we'd still add another type class, and it would be pretty similar as far as I can tell.

Reply:
Yeah, it just shifts us from needing a configuration when we call `pretty` vs when we convert it to a string, which might not be a big deal.